[CCLEX-175]: added cluster events page

AntonNos opened the pull request:
### **Events table (events aren't empty):**
<img width="1792" alt="Screenshot 2022-12-04 at 20 16 13" src="https://user-images.githubusercontent.com/33761040/205508396-439ff9ad-a6ba-46c8-ac2e-91ebd2582423.png">

### **Light mode**
FYI: We also have that blue borders in other tables in Lens.
<img width="1792" alt="Screenshot 2022-12-06 at 15 30 35" src="https://user-images.githubusercontent.com/33761040/205925744-1c0ffe03-6de0-4d94-9227-54a80f338506.png">


### **Filtered table, items not found:**
<img width="1792" alt="Screenshot 2022-12-04 at 20 18 43" src="https://user-images.githubusercontent.com/33761040/205508420-f6d51c7c-18bd-4b77-a42a-d12ae8f62fd0.png">

### **Events table (events are empty):**
<img width="1792" alt="Screenshot 2022-12-04 at 20 18 54" src="https://user-images.githubusercontent.com/33761040/205508456-74549f25-48af-4076-bb04-3b447c788f1d.png">

### **Test coverage for `useTableSearch.js`:**
<img width="733" alt="Screenshot 2022-12-04 at 20 20 00" src="https://user-images.githubusercontent.com/33761040/205508475-0cba6b02-d2bd-4461-ae2f-b9069a7c23fc.png">

stefcameron commented:
👀

stefcameron reviewed (changes requested):
Nice work! There are a lot of comments, but most of them are hopefully simple variable renamings for make the code clearer, especially when coming back to it for some reason in a few months or a year and not remembering how it all works...

Also, what does it look like the Light theme?

john-papadami commented on src/renderer/components/ClusterPage/Events/EventsTable.js:
```diff
@@ -0,0 +1,220 @@
+  textAlign: 'center',
+}));
+
+const ResetFiltersButton = styled.button(() => ({
+  color: 'var(--textColorAccent)',
+  borderBottom: '1px dotted',
+}));
```
> 
I think it looks better as a white link, not intrusive, and uses to text to go along with the message "No items found". But does this clear the search field or clears the search filter and sets sources dropdown to "All sources"? In any case I probably would name it "Reset search" since the input field says search events and should perform the latter.

AntonNos replied:
Yes, I got this styles from similar table in Lens (Cluster page -> Events).
<img width="1254" alt="Screenshot 2022-12-06 at 13 52 12" src="https://user-images.githubusercontent.com/33761040/205904412-dc6db06c-139b-4a9f-9c7d-78511f10e339.png">

But yes @john-papadami , you are right. I checked behaviour of that in Lens table, and it clears only search. It's actually naming of that button caused me a misunderstanding with the functional. So yes, it would be better to rename it.
So should I keep current styles for it?

john-papadami replied:
@AntonNos Yes, let's keep the current style.

AntonNos replied:
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+        allValues = [];
+      }
+      for (let key in singleItem) {
+        if (typeof singleItem[key] === 'object') {
```
Makes sense , fixed. BTW, I added light theme screenshot to description.

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+export const useTableSearch = ({ searchQuery, data }) => {
+  const [searchedData, setSearchedData] = useState([]);
+  const [origData, setOrigData] = useState([]);
+  const [searchIndex, setSearchIndex] = useState([]);
```
Thanks, fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+
+      const searchInd = data.map((singleItem) => {
+        const allValues = getDataForSearch(singleItem);
+        return { allValues: allValues.toString() };
```
Makes sense , fixed.

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+    if (searchQuery) {
+      const reqData = searchIndex.map((singleItem, index) => {
+        if (
+          singleItem.allValues
+            .toLowerCase()
+            .indexOf(searchQuery.toLowerCase()) >= 0
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Returns data, filtered by search query.
+ * @param {string} searchQuery Search query.
+ * @param {Array} data Origin data to filter.
```
Yes, sure. It sounds better. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+ */
+export const useTableSearch = ({ searchQuery, data }) => {
+  const [searchedData, setSearchedData] = useState([]);
+  const [origData, setOrigData] = useState([]);
```
You are right, it works without it. I added it to have origin object, but yes, we already have it in parameters.🤷‍♂️
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+ * @returns {Array} Filtered data.
+ */
+export const useTableSearch = ({ searchQuery, data }) => {
+  const [searchedData, setSearchedData] = useState([]);
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+
+/**
+ * Returns data, filtered by search query.
+ * @param {string} searchQuery Search query.
```
Sounds better. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+        })
+      );
+    } else {
+      setSearchedData(origData);
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/useTableSearch.js:
```diff
@@ -0,0 +1,72 @@
+    }
+  }, [searchQuery, origData, searchIndex]);
+
+  return { searchedData };
```
Fixed

AntonNos commented on src/strings.ts:
```diff
@@ -422,6 +422,25 @@ export const clusterPage: Dict = {
+          machine: () => 'Machine',
+          count: () => 'Count',
+        },
+        emptyList: () => 'Item list is empty',
```
Fixed

AntonNos commented on src/strings.ts:
```diff
@@ -422,6 +422,25 @@ export const clusterPage: Dict = {
+          count: () => 'Count',
+        },
+        emptyList: () => 'Item list is empty',
+        noItemsFound: () => 'No items found.',
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/__tests__/useTableSearch.spec.js:
```diff
@@ -0,0 +1,56 @@
+
+const TestComponent = ({ searchQuery }) => {
+  const { searchedData } = useTableSearch({
+    searchQuery,
+    data: mockedData,
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsTable.js:
```diff
@@ -0,0 +1,220 @@
+      label: propTypes.string,
+    })
+  ).isRequired,
+  events: propTypes.array.isRequired,
```
Fixed

stefcameron replied:
`modifiedEvents` should just be an array containing a subset of items from `clusterEntity.spec.events`. Oooh, but those items aren't `ResourceEvent` instance anymore: They've been converted to plain objects. My bad on that one. So `propTypes.arrayOf(propTypes.object)` is probably enough in this case.

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsTable.js:
```diff
@@ -0,0 +1,220 @@
+    })
+  ).isRequired,
+  events: propTypes.array.isRequired,
+  handleSortChange: propTypes.func.isRequired,
+  handleResetFilters: propTypes.func.isRequired,
```
Thanks, will tale into account. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsTable.js:
```diff
@@ -0,0 +1,220 @@
+  },
+} = strings;
+
+const formatDate = (date) => {
+  if (!date) {
+    return undefined;
+  }
+
+  const dateObj = new Date(date);
+  return dateObj.getTime() === 0
+    ? undefined
+    : dayjs(date).format('YYYY-MM-DD, HH:mm:ss');
+};
```
We have a bit different format in Figma, thats why I added new internal `formatDate`.
`4 days ago (2022-12-02, 20:31:18)` -> `2022-11-27, 17:37:30`.
If we can use date format like in other tabs, I will change it. 
What do you think? @stefcameron @john-papadami . Can we use this variant (`4 days ago (2022-12-02, 20:31:18)`)?

AntonNos replied:
Actually, if we've already decided to use one format, let's keep it. I'll fix it in the future if it doesn't fit

AntonNos replied:
Actually, if we've already decided to use one format, let's keep it. I'll fix it in the future if it doesn't fit

stefcameron replied:
Just add a parameter to `rendererUtil.formatDate(date, includeRelative = true)` and then call `rendererUtil.formatDate(date, false)` from here to produce just the `'YYYY-MM-DD, HH:mm:ss'` part.

stefcameron replied:
(I'm sorry I didn't realize there was indeed a slight difference. But I think adding a parameter is better than creating another function because the `'YYYY-MM-DD, HH:mm:ss'` is the same between the two, and how we want to print an exact date).

I think if we show a relative date in the future, it should be in a separate column with a heading like "Age".

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+          <SyncButton
+            isDisabled={isFetching || status !== CONNECTION_STATUSES.CONNECTED}
+            isFetching={isFetching}
+            onClick={() => handleSync()}
```
Moved it outside of component

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+    );
+  };
+
+  const getSourceOptions = () => {
+    const uniqueSources = [
+      ...new Set(events.map(({ metadata: { source } }) => source)),
+    ];
+
+    return [
+      defaultSourceOption,
+      ...uniqueSources.map((source) => ({
+        value: source,
+        label: source,
+      })),
+    ];
+  };
```
Done

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+    };
+  }, [clouds, cloudUrl]);
+
+  const handleSync = () => {
+    IpcRenderer.getInstance().invoke(
+      consts.ipcEvents.invoke.SYNC_NOW,
+      clusterEntity.metadata.cloudUrl
+    );
+  };
```
Fixed for `handleSelectChange`, `handleSearchChange`, `handleSortChange` and `handleResetFilters`, but not for `handleSync`, because I moved it outside of component.

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+  const { clouds } = useClouds();
+
+  const [isLoading, setIsLoading] = useState(false);
+  const [isFetching, setIsFetching] = useState(false);
+  const [status, setStatus] = useState(clouds[cloudUrl].status);
```
Yes, sure. Done

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+  const [isFiltered, setIsFiltered] = useState(false);
+
+  const { searchedData } = useTableSearch({
+    searchQuery: filters.searchQuery,
+    data: events,
```
Done

AntonNos commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -0,0 +1,290 @@
+  const [isLoading, setIsLoading] = useState(false);
+  const [isFetching, setIsFetching] = useState(false);
+  const [status, setStatus] = useState(clouds[cloudUrl].status);
+  const [events, setEvents] = useState([]);
+  const [modifiedEvents, setModifiedEvents] = useState([]);
```
Fixed

stefcameron commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -222,64 +234,81 @@ export const EventsPanel = ({ clusterEntity }) => {
+              isCloudFetching || cloudStatus !== CONNECTION_STATUSES.CONNECTED
+            }
+            isCloudFetching={isCloudFetching}
+            onClick={() => handleSync(clusterEntity.metadata.cloudUrl)}
```
This is still generating a new handler function on every render...

AntonNos replied:
Fixed

stefcameron reviewed (changes requested):
Thank you for working through all my previous comments. I appreciate it! Just a few more...

stefcameron commented on src/renderer/components/ClusterPage/useTableSearch.js:
```diff
@@ -0,0 +1,67 @@
+          !Array.isArray(singleItem[key])
+        ) {
+          getDataForSearch(singleItem[key], allValues);
+        } else {
```
We also need to convert date objects using `rendererUtil.formatDate(date, false)` in the same way we display it because that's what the user sees and may search on.

AntonNos replied:
<img width="1792" alt="Screenshot 2022-12-08 at 14 48 08" src="https://user-images.githubusercontent.com/33761040/206450788-1f9d3297-afcc-4727-b408-8965e594531d.png">
<img width="1792" alt="Screenshot 2022-12-08 at 14 48 44" src="https://user-images.githubusercontent.com/33761040/206450775-624ba3dc-0085-44e2-9416-c4595a94a424.png">

AntonNos replied:
Without `!Array.isArray(singleItem[key])` is works well (I mean we don't have [object Object] in string). It converting all objects and arrays to strings with it values. I added some screenshots from JSbin where is a case that you expected.

stefcameron replied:
If think the array case just happens to kind of work because you're using a `for..in` loop to iterate properties of objects, and `typeof [] === 'object'`. So without the `!Array.isArray()` check, you determine the array is an object and you recurse, at which point you iterate its elements.

But the problem with `for..in` is that it it iterates over all enumerable properties of an object -- including those inherited via the prototype chain. I'm surprised more stuff isn't coming out than it is.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in

And since this function's purpose is to traverse a generic object without knowing its structure, it should be safely iterating over nested objects and arrays.

AntonNos replied:
Fixed it in way, that we discussed in Slack

stefcameron commented:
Thanks for posting the light theme screen shot. Why are there blue lines above and below the table header, only in light mode?

<img width="1223" alt="image" src="https://user-images.githubusercontent.com/2855350/206015931-c4e0ee4c-4531-4fe6-8888-8aa74511b1c6.png">

stefcameron commented:
There's strange behavior when sorting columns: Even when a column has identical values on all visible rows, the search order still changes (can't tell exactly based on what; maybe it's the date?) when I toggle asc vs desc. If all the values are the same, once the column has been sorted once (i.e. when it wasn't the primary sort), shouldn't it remain the same afterward?

![lex-cluster-events-panel](https://user-images.githubusercontent.com/2855350/206016417-7fe3b55f-d94f-45e9-818a-f507050cab86.gif)

AntonNos commented on src/renderer/components/ClusterPage/useTableSearch.js:
```diff
@@ -0,0 +1,67 @@
+ * Returns data, filtered by search query.
+ * @param {string} searchText Text to be searched for.
+ * @param {Array} searchItems Unfiltered list of items to search.
+ * @returns {Array} Filtered data.
```
Fixed

AntonNos commented:
> There's strange behavior when sorting columns: Even when a column has identical values on all visible rows, the search order still changes (can't tell exactly based on what; maybe it's the date?) when I toggle asc vs desc. If all the values are the same, once the column has been sorted once (i.e. when it wasn't the primary sort), shouldn't it remain the same afterward?

Fixed. It was because of this `filters.sort.isAsc ? sortedEvents : sortedEvents.reverse()`. I changed login and made it without `reverse()`

stefcameron commented on src/renderer/components/ClusterPage/Events/EventsPanel.js:
```diff
@@ -160,25 +160,37 @@ export const EventsPanel = ({ clusterEntity }) => {
-    setFilteredEvents(
-      filters.sort.isAsc ? sortedEvents : sortedEvents.reverse()
-    );
+    // setFilteredEvents(
+    //   filters.sort.isAsc ? sortedEvents : sortedEvents.reverse()
+    // );
```
Can be removed?

AntonNos replied:
Yes, done. Thanks

stefcameron commented on src/renderer/rendererUtil.js:
```diff
@@ -27,15 +27,17 @@ export const getLensClusters = function () {
  * @returns {string|undefined} If a date is given, then formatted. If falsy, or the date is the epoch, undefined.
  */
-export const formatDate = (date) => {
+export const formatDate = (date, includeRelative = true) => {
```
New parameter should be added to JSDoc

AntonNos replied:
Done

stefcameron commented on src/renderer/components/ClusterPage/useTableSearch.js:
```diff
@@ -1,10 +1,24 @@
+
+/**
+ * Check if a string is in ISO format
+ * @param {string} str Unfiltered list of items to search.
```
Is it really a __list__? Looks more like a piece of text to check for a date. 🤔

AntonNos replied:
Ohh, sorry, copied JSDoc from `useTableSearch`, but forgot to change parameter description.

stefcameron commented on src/renderer/components/ClusterPage/useTableSearch.js:
```diff
@@ -16,14 +30,16 @@ export const useTableSearch = ({ searchText, searchItems }) => {
-          allValues.push(`${singleItem[key]}`);
+          allValues.push(
+            `${
+              isIsoDate(singleItem[key])
+                ? formatDate(singleItem[key], false)
```
Could you add a short comment here stating that we're formatting in the same we assume dates are displayed so as to search text as the user see's it in the table? That would help clarify why we're doing this special case.

AntonNos replied:
Yes, sure, done.

stefcameron reviewed (approved):
I think it's good to go! I tested it locally and it works really well. Thank you for your work on this! 🙂
